README.md
# Notice

We have abandoned this package because Laravel 7 introduced native support for CORS. Only use this package if you're on Laravel 6 or below.

# Send CORS headers in a Laravel application

[![Latest Version on Packagist](https://img.shields.io/packagist/v/spatie/laravel-cors.svg?style=flat-square)](https://packagist.org/packages/spatie/laravel-cors)
[![Build Status](https://img.shields.io/travis/spatie/laravel-cors/master.svg?style=flat-square)](https://travis-ci.org/spatie/laravel-cors)
[![Quality Score](https://img.shields.io/scrutinizer/g/spatie/laravel-cors.svg?style=flat-square)](https://scrutinizer-ci.com/g/spatie/laravel-cors)
[![StyleCI](https://styleci.io/repos/113957368/shield?branch=master)](https://styleci.io/repos/113957368)
[![Total Downloads](https://img.shields.io/packagist/dt/spatie/laravel-cors.svg?style=flat-square)](https://packagist.org/packages/spatie/laravel-cors)

This package will add CORS headers to the responses of your Laravel or Lumen app. For more infomation about CORS, see the [Mozilla CORS documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS).

This package supports preflight requests and is easily configurable to fit your needs.

## Installation

- [Laravel](#laravel)
- [Lumen](#lumen)

### Laravel

You can install the package via Composer:

```bash
composer require spatie/laravel-cors
```

The package will automatically register its service provider.

The provided `Spatie\Cors\Cors` middleware must be registered in the global middleware group.

```php
// app/Http/Kernel.php

protected $middleware = [
    ...
    \Spatie\Cors\Cors::class
];
```

```php
php artisan vendor:publish --provider="Spatie\Cors\CorsServiceProvider" --tag="config"
```

This is the default content of the config file published at `config/cors.php`:

```php
return [
    /*
     * A cors profile determines which origins, methods, headers are allowed for
     * a given requests. The `DefaultProfile` reads its configuration from this
     * config file.
     *
     * You can easily create your own cors profile.
     * More info: https://github.com/spatie/laravel-cors/#creating-your-own-cors-profile
     */
    'cors_profile' => Spatie\Cors\CorsProfile\DefaultProfile::class,

    /*
     * This configuration is used by `DefaultProfile`.
     */
    'default_profile' => [

        'allow_credentials' => false,

        'allow_origins' => [
            '*',
        ],

        'allow_methods' => [
            'POST',
            'GET',
            'OPTIONS',
            'PUT',
            'PATCH',
            'DELETE',
        ],

        'allow_headers' => [
            'Content-Type',
            'X-Auth-Token',
            'Origin',
            'Authorization',
        ],

        'expose_headers' => [
            'Cache-Control',
            'Content-Language',
            'Content-Type',
            'Expires',
            'Last-Modified',
            'Pragma',
        ],

        'forbidden_response' => [
            'message' => 'Forbidden (cors).',
            'status' => 403,
        ],

        /*
         * Preflight request will respond with value for the max age header.
         */
        'max_age' => 60 * 60 * 24,
    ],
];
```

### Lumen

You can install the package via Composer:

```bash
composer require spatie/laravel-cors
```

Copy the config file from the vendor directory:

```bash
cp vendor/spatie/laravel-cors/config/cors.php config/cors.php
```

Register the config file, the middleware and the service provider in `bootstrap/app.php`:

```php
$app->configure('cors');

$app->middleware([
    Spatie\Cors\Cors::class,
]);

$app->register(Spatie\Cors\CorsServiceProvider::class);
```

## Usage

With the middleware installed your API routes should now get appropriate CORS headers. Preflight requests will be handled as well. If a request comes in that is not allowed, Laravel will return a `403` response.

The default configuration of this package allows all requests from any origin (denoted as `'*'`). You probably want to at least specify some origins relevant to your project. If you want to allow requests to come in from `https://spatie.be` and `https://laravel.com` add those domains to the config file:

```php
// config/cors.php

    ...
    'default_profile' => [

    'allow_origins' => [
        'https://spatie.be',
        'https://laravel.com',
    ],
    ...
...
```

If you, for example, want to allow all subdomains from a specific domain, you can use the wildcard asterisk (`*`) and specifiy that:
```php
// config/cors.php

    ...
    'default_profile' => [

    'allow_origins' => [
        'https://spatie.be',
        'https://laravel.com',

        'https://*.spatie.be',
        'https://*.laravel.com',
    ],
    ...
...
```

### Creating your own CORS profile

Imagine you want to specify allowed origins based on the user that is currently logged in. In that case the `DefaultProfile` which just reads the config file won't cut it. Fortunately it's very easy to write your own CORS profile, which is simply a class that extends `Spatie\Cors\DefaultProfile`.

Here's a quick example where it is assumed that you've already added an `allowed_domains` column on your user model:

```php
namespace App\Services\Cors;

use Spatie\Cors\CorsProfile\DefaultProfile;

class UserBasedCorsProfile extends DefaultProfile
{
    public function allowOrigins(): array
    {
        return Auth::user()->allowed_domains;
    }
}
```

You can override the default HTTP status code and message returned when a request is forbidden by editing the `forbidden_response` array in your configuration file:

```php
'forbidden_response' => [
    'message' => 'Your request failed',
    'status' => 400,
],
```

Don't forget to register your profile in the config file.

```php
// config/cors.php

 ...
 'cors_profile' => App\Services\Cors\UserBasedCorsProfile::class,
 ...
```

In the example above we've overwritten the `allowOrigins` method, but of course you may choose to override any of the methods present in `DefaultProfile`.

## Testing

``` bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security

If you discover any security related issues, please email freek@spatie.be instead of using the issue tracker.

## Alternatives

- [barryvdh/laravel-cors](https://github.com/barryvdh/laravel-cors): a tried and tested package. Our package is a modern rewrite of the basic features of Barry's excellent one. We created our own solution because we needed our configuration to be [very flexible](#creating-your-own-cors-profile).

## Postcardware

You're free to use this package, but if it makes it to your production environment we highly appreciate you sending us a postcard from your hometown, mentioning which of our package(s) you are using.

Our address is: Spatie, Samberstraat 69D, 2060 Antwerp, Belgium.

We publish all received postcards [on our company website](https://spatie.be/en/opensource/postcards).

## Credits

- [Freek Van der Herten](https://github.com/freekmurze)
- [All Contributors](../../contributors)

## Support us

Spatie is a webdesign agency based in Antwerp, Belgium. You'll find an overview of all our open source projects [on our website](https://spatie.be/opensource).

Does your business depend on our contributions? Reach out and support us on [Patreon](https://www.patreon.com/spatie).
All pledges will be dedicated to allocating workforce on maintenance and new awesome stuff.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-cors` will be documented in this file

## 1.7.0 - 2021-06-10

- add PHP 8.0 support

## 1.6.0 - 2019-09-04

- make compatible with Laravel 6

## 1.5.2 - 2019-07-19

- improve responses for preflight requests

## 1.5.1 - 2019-06-06

- change preflight response status code to 204

## 1.5.0 - 2019-04-17

- add wildcard matching for subdomains (#51)

## 1.4.0 - 2018-02-27

- drop support for PHP 7.1

## 1.3.2 - 2018-01-28

- fix for Lumen

## 1.3.1 - 2018-08-31

- return default value for allow credentials when not available in config

## 1.3.0 - 2018-08-30

**THIS VERSION DOES NOT WORK, DO NOT USE**

- add support for `Access-Control-Allow-Credentials`

## 1.2.3 - 2018-08-27

- add support for Laravel 5.7

## 1.2.2 - 2018-05-05

- add support for streamed responses

## 1.2.1 - 2018-04-24

- add default value for `expose_headers` config option

## 1.2.0 - 2018-03-30

- add an option to manage `Access-Control-Expose-Headers`

## 1.1.0 - 2018-03-09

- allow to customize error message

## 1.0.3 - 2018-02-28

- fix typo in config

## 1.0.2 - 2018-02-12

- add support for Lumen

## 1.0.1 - 2018-02-07
    
- add support for Laravel 5.6

## 1.0.0 - 2017-12-19

- initial release


tests/CorsTest.php
<?php

namespace Spatie\Cors\Tests;

class CorsTest extends TestCase
{
    /** @test */
    public function it_adds_the_cors_headers_on_a_valid_requests()
    {
        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Allow-Origin', '*')
            ->assertSee('real content');
    }

    /** @test */
    public function it_adds_the_wildcard_in_the_cors_headers_on_a_valid_request_if_no_allow_origins_are_set()
    {
        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Allow-Origin', '*')
            ->assertSee('real content');
    }

    /** @test */
    public function it_adds_the_credentials_cors_headers_on_a_valid_request_if_allow_credentials_is_set_to_true()
    {
        config()->set('cors.default_profile.allow_credentials', true);

        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Allow-Credentials', 'true')
            ->assertHeader('Access-Control-Allow-Origin', 'https://spatie.be')
            ->assertSee('real content');
    }

    /** @test */
    public function it_does_not_add_the_credentials_cors_headers_on_a_valid_request_if_allow_credentials_is_set_to_false()
    {
        config()->set('cors.default_profile.allow_credentials', false);

        $response = $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Allow-Origin', '*')
            ->assertSee('real content');

        $headerName = 'Access-Control-Allow-Credentials';

        $this->assertFalse($response->headers->has($headerName), "Unexpected header [{$headerName}] is present on response.");
    }

    /** @test */
    public function it_adds_the_origin_domain_in_the_cors_headers_on_a_valid_request()
    {
        config()->set('cors.default_profile.allow_origins', [
            'https://spatie.be',
            'https://laravel.com',
        ]);

        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Allow-Origin', 'https://spatie.be')
            ->assertSee('real content');
    }

    /** @test */
    public function it_adds_the_origin_domain_in_the_cors_headers_on_a_valid_request_with_wildcard_content()
    {
        config()->set('cors.default_profile.allow_origins', [
            'https://*.be',
            'https://*.com',
        ]);

        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Allow-Origin', 'https://spatie.be')
            ->assertSee('real content');
    }

    /** @test */
    public function it_throws_error_on_an_invalid_request_with_wildcard_content()
    {
        config()->set('cors.default_profile.allow_origins', [
            'https://*.spatie.be',
            'https://*.com',
        ]);

        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertStatus(403)
            ->assertSee('Forbidden (cors).');
    }

    /** @test */
    public function it_adds_the_allowed_expose_headers_in_the_cors_headers_on_a_valid_request()
    {
        config()->set('cors.default_profile.expose_headers', [
            'Authorization',
            'X-Foo-Header',
        ]);

        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Expose-Headers', 'Authorization, X-Foo-Header')
            ->assertHeader('Access-Control-Allow-Origin', '*')
            ->assertSee('real content');
    }

    /** @test */
    public function it_will_send_a_403_for_invalid_requests()
    {
        config()->set('cors.default_profile.allow_origins', ['https://spatie.be']);

        $this
            ->sendRequest('POST', 'https://laravel.com')
            ->assertStatus(403)
            ->assertSee('Forbidden (cors).');
    }

    /** @test */
    public function it_sends_the_custom_forbidden_response_for_invalid_requests()
    {
        $forbiddenMessage = 'Custom forbidden message';
        $forbiddenStatus = 400;

        config()->set('cors.default_profile.allow_origins', ['https://spatie.be']);
        config()->set('cors.default_profile.forbidden_response.message', $forbiddenMessage);
        config()->set('cors.default_profile.forbidden_response.status', $forbiddenStatus);

        $this
            ->sendRequest('POST', 'https://laravel.com')
            ->assertStatus($forbiddenStatus)
            ->assertSee($forbiddenMessage);
    }

    /** @test */
    public function it_will_be_a_valid_profile_if_expose_header_is_not_set()
    {
        config()->set('cors.default_profile.expose_headers', null);

        $this
            ->sendRequest('POST', 'https://spatie.be')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Expose-Headers', '')
            ->assertHeader('Access-Control-Allow-Origin', '*')
            ->assertSee('real content');
    }

    /** @test */
    public function it_adds_the_cors_headers_on_a_valid_stream_request()
    {
        $this
            ->sendRequest('POST', 'https://spatie.be', 'test-cors-stream')
            ->assertSuccessful()
            ->assertHeader('Access-Control-Allow-Origin', '*');
    }

    public function sendRequest(string $method, string $origin, string $uri = 'test-cors')
    {
        $headers = [
            'Origin' => $origin,
        ];

        $server = $this->transformHeadersToServerVars($headers);

        return $this->call($method, $uri, [], [], [], $server);
    }
}


tests/TestCase.php
<?php

namespace Spatie\Cors\Tests;

use Illuminate\Contracts\Http\Kernel;
use Illuminate\Support\Facades\Route;
use Orchestra\Testbench\TestCase as Orchestra;
use Spatie\Cors\Cors;
use Symfony\Component\HttpFoundation\StreamedResponse;

class TestCase extends Orchestra
{
    public function setUp(): void
    {
        parent::setUp();

        $this->setupRoutes();
    }

    protected function setupRoutes()
    {
        Route::post('test-cors', function () {
            return 'real content';
        });

        Route::post('test-cors-stream', function () {
            return new StreamedResponse();
        });
    }

    /**
     * @param \Illuminate\Foundation\Application $app
     */
    protected function getEnvironmentSetUp($app)
    {
        $app->make(Kernel::class)->prependMiddleware(Cors::class);
    }

    /**
     * @param \Illuminate\Foundation\Application $app
     *
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [
            \Spatie\Cors\CorsServiceProvider::class,
        ];
    }
}


tests/ConfigTest.php
<?php

namespace Spatie\Cors\Tests;

use Spatie\Cors\CorsServiceProvider;
use Spatie\Cors\Exceptions\InvalidCorsProfile;

class ConfigTest extends TestCase
{
    /** @test */
    public function it_will_throw_an_exception_when_an_invalid_profile_is_set()
    {
        config()->set('cors.cors_profile', 'FakeClass');

        $this->expectException(InvalidCorsProfile::class);

        (new CorsServiceProvider($this->app))->boot();
    }
}


src/Cors.php
<?php

namespace Spatie\Cors;

use Closure;
use Spatie\Cors\CorsProfile\CorsProfile;

class Cors
{
    /** @var \Spatie\Cors\CorsProfile\CorsProfile */
    protected $corsProfile;

    public function __construct(CorsProfile $corsProfile)
    {
        $this->corsProfile = $corsProfile;
    }

    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle($request, Closure $next)
    {
        if (! $this->isCorsRequest($request)) {
            return $next($request);
        }

        $this->corsProfile->setRequest($request);

        if (! $this->corsProfile->isAllowed()) {
            return $this->forbiddenResponse();
        }

        if ($this->isPreflightRequest($request)) {
            return $this->handlePreflightRequest();
        }

        $response = $next($request);

        return $this->corsProfile->addCorsHeaders($response);
    }

    protected function isCorsRequest($request): bool
    {
        if (! $request->headers->has('Origin')) {
            return false;
        }

        return $request->headers->get('Origin') !== $request->getSchemeAndHttpHost();
    }

    protected function isPreflightRequest($request): bool
    {
        return $request->getMethod() === 'OPTIONS';
    }

    protected function handlePreflightRequest()
    {
        if (! $this->corsProfile->isAllowed()) {
            return $this->forbiddenResponse();
        }

        return $this->corsProfile->addPreflightHeaders(response(null, 204));
    }

    protected function forbiddenResponse()
    {
        $message = config('cors.default_profile.forbidden_response.message');
        $status = config('cors.default_profile.forbidden_response.status');

        return response(
            $message ?? 'Forbidden (cors).',
            $status ?? 403
        );
    }
}


src/CorsServiceProvider.php
<?php

namespace Spatie\Cors;

use Illuminate\Support\ServiceProvider;
use Spatie\Cors\CorsProfile\CorsProfile;
use Spatie\Cors\CorsProfile\DefaultProfile;
use Spatie\Cors\Exceptions\InvalidCorsProfile;

class CorsServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if ($this->isLaravel() && $this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/cors.php' => config_path('cors.php'),
            ], 'config');
        }

        $configuredCorsProfile = config('cors.cors_profile');

        if (! is_a($configuredCorsProfile, DefaultProfile::class, true)) {
            throw InvalidCorsProfile::profileDoesNotExtendDefaultProfile($configuredCorsProfile);
        }

        $this->app->bind(CorsProfile::class, $configuredCorsProfile);
    }

    public function register()
    {
        if ($this->isLaravel()) {
            $this->mergeConfigFrom(__DIR__.'/../config/cors.php', 'cors');
        }
    }

    protected function isLaravel()
    {
        return ! preg_match('/lumen/i', app()->version());
    }
}


src/CorsLiteServiceProvider.php
<?php

namespace Spatie\Cors;

use Illuminate\Support\ServiceProvider;
use Spatie\Cors\CorsProfile\CorsProfile;
use Spatie\Cors\CorsProfile\DefaultProfile;
use Spatie\Cors\Exceptions\InvalidCorsProfile;

class CorsServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/cors.php' => config_path('cors.php'),
            ], 'config');
        }

        $configuredCorsProfile = config('cors.cors_profile');

        if (! is_a($configuredCorsProfile, DefaultProfile::class, true)) {
            throw InvalidCorsProfile::profileDoesNotExtendDefaultProfile($configuredCorsProfile);
        }

        $this->app->bind(CorsProfile::class, $configuredCorsProfile);
    }

    public function register()
    {
        $this->mergeConfigFrom(__DIR__.'/../config/cors.php', 'cors');
    }
}


src/CorsProfile/DefaultProfile.php
<?php

namespace Spatie\Cors\CorsProfile;

class DefaultProfile implements CorsProfile
{
    /** @var \Illuminate\Http\Request */
    protected $request;

    public function setRequest($request)
    {
        $this->request = $request;
    }

    public function allowCredentials(): bool
    {
        return config('cors.default_profile.allow_credentials') ?? false;
    }

    public function allowOrigins(): array
    {
        return config('cors.default_profile.allow_origins');
    }

    public function allowMethods(): array
    {
        return config('cors.default_profile.allow_methods');
    }

    public function allowHeaders(): array
    {
        return config('cors.default_profile.allow_headers');
    }

    public function exposeHeaders(): array
    {
        return config('cors.default_profile.expose_headers') ?? [];
    }

    public function maxAge(): int
    {
        return config('cors.default_profile.max_age');
    }

    public function addCorsHeaders($response)
    {
        if ($this->allowCredentials()) {
            $response->headers->set('Access-Control-Allow-Credentials', 'true');
        }

        $response->headers->set('Access-Control-Allow-Origin', $this->allowedOriginsToString());
        $response->headers->set('Access-Control-Expose-Headers', $this->toString($this->exposeHeaders()));

        return $response;
    }

    public function addPreflightHeaders($response)
    {
        if ($this->allowCredentials()) {
            $response->headers->set('Access-Control-Allow-Credentials', 'true');
        }

        $response->headers->set('Access-Control-Allow-Methods', $this->toString($this->allowMethods()));
        $response->headers->set('Access-Control-Allow-Headers', $this->toString($this->allowHeaders()));
        $response->headers->set('Access-Control-Allow-Origin', $this->allowedOriginsToString());
        $response->headers->set('Access-Control-Max-Age', $this->maxAge());

        return $response;
    }

    public function isAllowed(): bool
    {
        if (! in_array($this->request->method(), $this->allowMethods())) {
            return false;
        }

        if (in_array('*', $this->allowOrigins())) {
            return true;
        }

        $matches = collect($this->allowOrigins())->filter(function ($allowedOrigin) {
            return fnmatch($allowedOrigin, $this->request->header('Origin'));
        });

        return $matches->count() > 0;
    }

    protected function toString(array $array): string
    {
        return implode(', ', $array);
    }

    protected function allowedOriginsToString(): string
    {
        if (! $this->isAllowed()) {
            return '';
        }

        if (in_array('*', $this->allowOrigins()) && ! $this->allowCredentials()) {
            return '*';
        }

        return $this->request->header('Origin');
    }
}


config/cors.php
<?php

return [

    /*
     * A cors profile determines which origins, methods, headers are allowed for
     * a given requests. The `DefaultProfile` reads its configuration from this
     * config file.
     *
     * You can easily create your own cors profile.
     * More info: https://github.com/spatie/laravel-cors/#creating-your-own-cors-profile
     */
    'cors_profile' => Spatie\Cors\CorsProfile\DefaultProfile::class,

    /*
     * This configuration is used by `DefaultProfile`.
     */
    'default_profile' => [

        'allow_credentials' => false,

        'allow_origins' => [
            '*',
        ],

        'allow_methods' => [
            'POST',
            'GET',
            'OPTIONS',
            'PUT',
            'PATCH',
            'DELETE',
        ],

        'allow_headers' => [
            'Content-Type',
            'X-Auth-Token',
            'Origin',
            'Authorization',
        ],

        'expose_headers' => [
            'Cache-Control',
            'Content-Language',
            'Content-Type',
            'Expires',
            'Last-Modified',
            'Pragma',
        ],

        'forbidden_response' => [
            'message' => 'Forbidden (cors).',
            'status' => 403,
        ],

        /*
         * Preflight request will respond with value for the max age header.
         */
        'max_age' => 60 * 60 * 24,
    ],
];
